crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,198 @@
     }
 
-    pub fn diagnostics(&mut self, url: &Url) -> Vec<Diagnostic> {
-        let Some(document) = self.document_by_url(url) else {
+    ///Tries to get the latest document from analysis.
+    ///Gives up and returns none after 5 seconds.
+    async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
+        let duration = std::time::Duration::from_secs(5);
+        tokio::time::timeout(duration, async {
+            //TODO: This should really be a condvar that is triggered by the latest being ready, this will do for now though
+            loop {
+                let docs = self.documents.lock().await;
+                if let Some(a) = docs.get(url) {
+                    if let Some(a) = a.latest_document.get() {
+                        return a.clone();
+                    }
+                }
+                drop(docs);
+                tokio::task::yield_now().await;
+            }
+        })
+        .await
+        .ok()
+    }
+
+    pub async fn diagnostics(&self, url: &Url) -> Vec<Diagnostic> {
+        let Some(document) = self.latest_document_by_url(url).await else {
             return vec![];
         };
         document.diagnostics()
     }
 
-    pub fn hover(&mut self, url: &Url, position: Position) -> Option<Hover> {
-        self.document_by_url(url)?.hover(position)
+    pub async fn hover(&self, url: &Url, position: Position) -> Option<Hover> {
+        self.latest_document_by_url(url).await?.hover(position)
     }
 
-    pub fn goto_definition(
-        &mut self,
+    pub async fn goto_definition(
+        &self,
         url: &Url,
         position: Position,
     ) -> Option<GotoDefinitionResponse> {
-        let document = self.document_by_url(url)?;
+        let document = self.latest_document_by_url(url).await?;
         let symbol = document.symbol_at(position)?;
         let def_document_url = document.module_url(symbol.module_id())?;
-        let def_document = self.document_by_url(&def_document_url)?;
+        let def_document = self.latest_document_by_url(&def_document_url).await?;
         def_document.definition(symbol)
     }
 
-    pub fn formatting(&mut self, url: &Url) -> Option<Vec<TextEdit>> {
-        let document = self.document_by_url(url)?;
+    pub async fn formatting(&self, url: &Url) -> Option<Vec<TextEdit>> {
+        let document = self.document_info_by_url(url).await?;
         document.format()
     }
 
-    pub fn semantic_tokens(&mut self, url: &Url) -> Option<SemanticTokensResult> {
-        let document = self.document_by_url(url)?;
+    pub async fn semantic_tokens(&self, url: &Url) -> Option<SemanticTokensResult> {
+        let document = self.document_info_by_url(url).await?;
         document.semantic_tokens()
     }
+    pub async fn completion_items(
+        &self,
+        url: &Url,
+        position: Position,
+    ) -> Option<CompletionResponse> {
+        trace!("starting completion ");
```

Comment:
```suggestion
        trace!("starting completion");
```

Maybe log the url and position